### ISSUE TITLE
Saving the Distilled Model

### ISSUE THREAD

**Issue by gadhane** [clovaai/overhaul-distillation#38]
When we want to save the trained distilled model, should we save the d_net or the s_net? And if it is the d_net how do we load the trained model later if we want to use it for prediction?
For example for s_net and t_net we can load as follows:

```
Teacher_path = 'WRN28-4_21.09.pt'
Student_path = 'WRNStudent.pt'

model1 = WRN.WideResNet(depth=28, widen_factor=4, num_classes=100)
state = torch.load(Teacher_path, map_location={'cuda:0': 'cpu'})['model']
model1.load_state_dict(state)

model2 = WRN.WideResNet(depth=16, widen_factor=4, num_classes=100)
state = torch.load(Student_path, map_location={'cuda:0': 'cpu'})
model2.load_state_dict(state)
```
assuming that the distilled model is d_net, how do we load it?

**Comment by bhheo** [comment]
Hi

You can simply save `s_net` to use it for prediction.
Because member variables `d_net.t_net` and `d_net.s_net` are links to the original network, you don't need to save `d_net` for inference.

https://github.com/clovaai/overhaul-distillation/blob/76344a84a7ce23c894f41a2e05b866c9b73fd85a/CIFAR-100/train_with_distillation.py#L147

**Comment by gadhane** [comment]
@bhheo Thank you for the reply. Additional question, why do we set the `t_net `in train mode at line 75?
[    d_net.t_net.train() ](https://github.com/clovaai/overhaul-distillation/blob/76344a84a7ce23c894f41a2e05b866c9b73fd85a/CIFAR-100/train_with_distillation.py#L75)

When we update the weights at line 93:
[        loss.backward() ](https://github.com/clovaai/overhaul-distillation/blob/76344a84a7ce23c894f41a2e05b866c9b73fd85a/CIFAR-100/train_with_distillation.py#L93), I guess it is updating both the `s_net `and `d_net `models. If we save the `t_net`, does this mean it is a teacher model trained with the distillation and we can use `t_net` for inferences?

Thank You.

**Comment by bhheo** [comment]
@GeryAdhane 
You can find the reason in Table 8 of my paper.
I used training mode because it showed better performance than eval mode.

My code doesn't train the teacher `t_net`.
`d_net.t_net.train()` is inherited from `nn.module.train()`.
In PyTorch library, `nn.module.train()` doesn't mean training.

This function simply assigns `self.training=True` for it and its submodules
https://github.com/pytorch/pytorch/blob/396c3b1d88d7624938a2bb0b287f2a19f1e89bb4/torch/nn/modules/module.py#L2148-L2168
Some specific modules such as BatchNorm use this flag `self.training` to determine the mode of module.
https://github.com/pytorch/pytorch/blob/396c3b1d88d7624938a2bb0b287f2a19f1e89bb4/torch/nn/modules/batchnorm.py#L148-L183
So, I want to use the teacher's batch-norm in training mode. That's all.
It doesn't train `t_net` and `t_net` do noting for inference.

**Comment by gadhane** [comment]
@bhheo 
Thank you for the clear explanation. Now I am able to trace and fully understand the code. 
Excellent work.

Thank you.
